### PR TITLE
agdaPackages.standard-library: 1.4 -> 1.5

### PR DESCRIPTION
**Pull request by iblech**
###### Motivation for this change

Just two hours ago the new version of the Agda standard library has been released, containg lots of new goodies.

This pull request bumps our package to the new version. I'm currently compiling the new version and will report back when it has finished.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by iblech** [comment]
The test run succeeded just fine! :-)

However, I noticed a problem with the categories library. After upgrading to stdlib 1.5, its compilation fails with:

```
 Checking Categories.Adjoint.Instance.0-Truncation (/build/source/src/Categories/Adjoint/Instance/0-Truncation.agda).
  Checking Categories.Category.Construction.0-Groupoid (/build/source/src/Categories/Category/Construction/0-Groupoid.agda). 
  Checking Categories.Category.Instance.Groupoids (/build/source/src/Categories/Category/Instance/Groupoids.agda).
  Checking Categories.Functor.Instance.0-Truncation (/build/source/src/Categories/Functor/Instance/0-Truncation.agda).
 Checking Categories.Adjoint.Instance.01-Truncation (/build/source/src/Categories/Adjoint/Instance/01-Truncation.agda).
  Checking Relation.Binary.OrderMorphism (/nix/store/0wbmxylcp4mjm40ghj6rf7d1lhafly9s-standard-library-1.5/src/Relation/Binary/OrderMorphism.agda).
Failed to write interface /nix/store/0wbmxylcp4mjm40ghj6rf7d1lhafly9s-standard-library-1.5/src/Relation/Binary/OrderMorphism.agdai.
/build/source/src/Categories/Adjoint/Instance/01-Truncation.agda:11,1-60
/nix/store/0wbmxylcp4mjm40ghj6rf7d1lhafly9s-standard-library-1.5/src/Relation/Binary/OrderMorphism.agdai:
openBinaryFile: permission denied (Permission denied)
builder for '/nix/store/98q18lp471h0mzj79afxqspz82i1375d-agda-categories-0.1.4.drv' failed with exit code 1
cannot build derivation '/nix/store/ippc4pcj16qmdsa879qb9f7cp63mgi7a-libraries.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/3crj84m1avs17ljgsm4v6nby16y0272m-agdaWithPackages-2.6.1.2.drv': 1 dependencies couldn't be built
error: build of '/nix/store/3crj84m1avs17ljgsm4v6nby16y0272m-agdaWithPackages-2.6.1.2.drv' failed
```

You should be able to reproduce this with

    $ nix-shell -I nixpkgs=https://github.com/iblech/nixpkgs/archive/patch-agdastdlib15.tar.gz -p "agda.withPackages (p: [ p.standard-library p.agda-categories ])"

I will try to look into that tomorrow (but can't promise).

**Comment by turion** [comment]
> **Edit:** Aha. This is because Relation.Binary.OrderMorphism has been deprecated and removed from std-lib's `Everything.agda`. As a consequence, we are no longer precompiling it; but on-demand compilation triggered by the categories library fails because of missing write permissions.

Yes, I agree that this is what happens.

> **Edit²:** I propose we upgrade standard-library to 1.5 (hence merge this pull request), but then also submit a pull request to upstream agda-categories, rendering it compatible with the new standard-library, and then update our package for agda-categories. What do you think?

I think it's better to have it the other way around. The aim is to always have a working agda library set. Every package in the following set should build: `lib.filterAttrs (name: value: (builtins.isAttrs value && builtins.hasAttr "isAgdaDerivation" value && value.isAgdaDerivation && !value.meta.broken)) agdaPackages`

So the PR to `agda-categories` should go first. Maybe they'll merge it very quickly, maybe they'll even release a new version soon. If this happens within a few days, update that as well, and we can merge both together. If it takes longer, we can temporarily point nixpkgs to the fixed fork.

**Comment by alexarice** [comment]
I would vote for holding off on this until agda-categories is ready

**Comment by turion** [comment]
You can use https://github.com/NixOS/nixpkgs/issues/98209#issuecomment-767482508 to check whether all agda libraries build

**Comment by turion** [comment]
@iblech Also, thanks a lot for helping to keep the agda package set up to date :)

**Comment by turion** [comment]
Upstream issue: https://github.com/agda/agda-categories/issues/242

**Comment by turion** [comment]
3 libraries are incompatible with this update :hand_over_mouth: let's see, hopefully they will all update soon

**Comment by MatthewDaggitt** [comment]
> Aha. This is because Relation.Binary.OrderMorphism has been deprecated and removed from std-lib's Everything.agda. As a consequence, we are no longer precompiling it; but on-demand compilation triggered by the categories library fails because of missing write permissions.

And after all that effort we went to have graceful deprecation of modules :cry: Sounds like we shouldn't remove the deprecated modules from `Everything.agda`?

**Comment by alexarice** [comment]
> 
> And after all that effort we went to have graceful deprecation of modules cry Sounds like we shouldn't remove the deprecated modules from `Everything.agda`?

This would keep stuff working but the reality is that every downstream library needs a new release every time standard library updates as things shouldn't really be using deprecated modules

**Comment by turion** [comment]
@MatthewDaggitt It's great that you put effort into graceful deprecation. The real problem is that Agda has no clear concept of what a package consists of, so we just build everything in `Everything.agda`, and nothing else. If people would make a local copy of `agda-stdlib`, they could still use the deprecated module, it would then be built when it's used. But `nixpkgs` tries to be "cleaner" than that and fail if anyone uses a deprecated module.

**Comment by alexarice** [comment]
Just to add to this, this is caused by nix packages being immutable once built, so if something doesn't get compiled when we build the package then it is unusable

**Comment by iblech** [comment]
Thank you all for your comments! I didn't know about our policy of always having a working Agda library set, but I totally agree that it is a sensible goal :-) Then let's hold off for now. I'll try to look into the upstream issues!

**Comment by turion** [comment]
>  I didn't know about our policy of always having a working Agda library set

It's not really set in stone, we're all in the process of figuring this out. My aim is that nixpkgs plays a similar role for Agda like stackage does for Haskell.

>  I'll try to look into the upstream issues!

That definitely a good idea. Often, a short PR is all it needs.

**Comment by ryanorendorff** [comment]
@turion I added `v0.2` tag for the `functional-linear-algebra` package that should work with agda stdlib 1.5. Should I make a PR on top of this one to add that new commit?

**Comment by turion** [comment]
@ryanorendorff Thanks a lot :) if it's a simple version bump then it's probably easiest if @iblech adds one commit to this PR here. If it's more, e.g. changed dependencies or build process, then it is probably helpful to make a PR onto the `iblech:patch-agdastdlib15` branch.

**Comment by iblech** [comment]
I'll quickly add the version bump to this pull request, and I have also given you two (turion and ryanorendorff) write access to my fork. Feel free to commit as you wish there. :-)

**Comment by turion** [comment]
Do all libraries build now? See https://github.com/NixOS/nixpkgs/issues/98209#issuecomment-767482508

**Comment by turion** [comment]
Huh. For some reason, ofborg builds some libraries. Not sure what's going on there.

**Comment by iblech** [comment]
Yes, they do! :-)

    $ nix-shell -I nixpkgs=https://github.com/iblech/nixpkgs/archive/patch-agdastdlib15.tar.gz -p 'agda.withPackages (lib.attrValues (lib.filterAttrs (name: value: (builtins.isAttrs value && builtins.hasAttr "isAgdaDerivation" value && value.isAgdaDerivation && !value.meta.broken)) agdaPackages))'
    unpacking 'https://github.com/iblech/nixpkgs/archive/patch-agdastdlib15.tar.gz'...
    [nix-shell:~] $

File: pkgs/development/libraries/agda/generic/default.nix
Comment (turion) [inline pkgs/development/libraries/agda/generic/default.nix:9]: Not sure why these lines were permuted.

Comment (turion) [reply pkgs/development/libraries/agda/generic/default.nix:9]: (But it doesn't matter.)

**Review by turion — APPROVED** [review]

File: pkgs/development/libraries/agda/generic/default.nix
Comment (iblech) [reply pkgs/development/libraries/agda/generic/default.nix:9]: I did it on purpose, because I was of the impression that we usually list the repo owner first and the repo name second. :-)

Comment (turion) [reply pkgs/development/libraries/agda/generic/default.nix:9]: Yes good point. That way, it's easier to parse it into github.com/effectfully/generic.

**Comment by iblech** [comment]
Awesome @AndersonTorres, thank you very much for the quick merge! :-) I really like that we have such a fast turnaround time. Thanks also to all the other involved persons!

**Comment by turion** [comment]
Yes, updating the agda standard library is a surprisingly pleasant process :)

**Pull request merged**
